### PR TITLE
[Feature] Add sample, Use single precision by default

### LINKED ISSUE #18: [Feature] Use single precision by default
@dominikandreasseitz If you could add a description.

### PR DESCRIPTION
**Pull request by dominikandreasseitz**
Closes #18

### CODE REVIEW

**Comment by Roland-djee** [comment]
Hey @dominikandreasseitz, I'm working on fixing this PR. One thing that came up is that in `test_hamevo_general` while constructing the Hamiltonian:

```python
(Pdb) hamevo
_HamiltonianEvolution(generator_name='I', target=((0, 1),), control=((None,),))
```

Just wondering what is the rationale behind the setting of `target` and `control`. What I understand is that: 

```python
    @staticmethod
    def parse_idx(
        idx: Tuple,
    ) -> Tuple:
        if isinstance(idx, (int, np.int64)):
            return ((idx,),)
        elif isinstance(idx, tuple):
            return (idx,)
        else:
            return (idx.astype(int),)
```

which makes `target` and `control` being tuples of tuples. Is this what is really intended ? For instance, defaults for `control` are `(None,)` which seems contradictory with the setting in `parse_idx`. Any guidance on this ? Thanks a lot :)

**Comment by dominikandreasseitz** [comment]
Hi @Roland-djee , the reason for the tuple of tuples design choice lies in the apply_gate function which can be either applied to a single gate or an arbitrary number of gates. To do so, the operators and supports are collected in a single tuple and then get applied in a single lambda expression. The "collection" part assumes a tuple of tuples for supports so thats the story behind that. I get its a bit counterintuitive but its part of legacy horqrux so i was hesitant to change it since the function is quite useful

**Comment by chMoussa** [comment]
Hey @RolandMacDoland I think I managed to fix this MR. What do you think?

**Comment by RolandMacDoland** [comment]
> Hey @RolandMacDoland I think I managed to fix this MR. What do you think?

Oh, wow cool thanks. Please feel free to request a review :)

File: horqrux/circuit.py [hunk 1]
```diff
@@ -1,47 +1,51 @@
...
 
     @property
     def param_names(self) -> list[str]:
-        return [str(op.param) for op in self.ansatz if isinstance(op, Parametric)]
+        return [
+            str(op.param)
+            for op in self.operations
+            if isinstance(op, Parametric)
+            and isinstance(op.param, str)
+            and op.param.startswith("v_")
```
Comment (RolandMacDoland) [inline horqrux/circuit.py:40]: Are these two checks necessary ?

File: horqrux/circuit.py [hunk 2]
```diff
@@ -58,11 +62,56 @@ def hea(n_qubits: int, n_layers: int, rot_fns: list[Callable] = [RX, RY, RX]) ->
     for _ in range(n_layers):
         for i in range(n_qubits):
             ops = [
-                fn(str(uuid4()), qubit)
+                fn("v_" + str(uuid4()), qubit)
```
Comment (RolandMacDoland) [inline horqrux/circuit.py:65]: Why doing this ?

File: horqrux/matrices.py [hunk 3]
```diff
@@ -23,11 +20,11 @@
         [0, 0.5 * (1 - 1j), 0.5 * (1 + 1j), 0],
         [0, 0, 0, 1],
     ],
-    dtype=jnp.complex128,
+    dtype=jnp.complex64,
```
Comment (RolandMacDoland) [inline horqrux/matrices.py:23]: hm isn't default in PyQ `complex128` ?

File: tests/test_adjoint.py [hunk 4]
```diff
@@ -26,13 +26,10 @@ def test_gradcheck() -> None:
     }
     state = random_state(MAX_QUBITS)
 
-    def adjoint_expfn(values) -> Array:
-        return adjoint_expectation(state, ops, observable, values)[0]
+    def exp_fn(values: dict, diff_mode: DiffMode) -> Array:
```
Comment (RolandMacDoland) [inline tests/test_adjoint.py:29]: Shouldn't `diff_mode` be defaulted ?

**Review by RolandMacDoland — COMMENTED** [review]
Thanks @chMoussa. Looking good. Few comments for the sake of discussion.

File: horqrux/circuit.py (hunk 1 above)
Comment (chMoussa) [reply horqrux/circuit.py:40]: Hm I guess what was left back then was a way to differentiate between variational parameters and non variational parameters. Let me examine this again later.

File: horqrux/circuit.py (hunk 2 above)
Comment (chMoussa) [reply horqrux/circuit.py:65]: no clue yet 😄

File: horqrux/matrices.py (hunk 3 above)
Comment (chMoussa) [reply horqrux/matrices.py:23]: Yes, not sure why was changed by ds

File: tests/test_adjoint.py (hunk 4 above)
Comment (chMoussa) [reply tests/test_adjoint.py:29]: I think the `expectation` function has a default. For the test, it's fine to keep it like this as we compare the `adjoint` with the `ad` diff

File: horqrux/circuit.py (hunk 2 above)
Comment (chMoussa) [reply horqrux/circuit.py:65]: @RolandMacDoland eventually, it's a convention for variational vs non variational parameters. I think I will change this a bit.

File: tests/test_adjoint.py (hunk 4 above)
Comment (chMoussa) [reply tests/test_adjoint.py]: Ok we fixed it in the test.

File: horqrux/circuit.py (hunk 2 above)
Comment (chMoussa) [reply horqrux/circuit.py]: @RolandMacDoland changed this by adding an extra field to specify feature map parameters from variational ones.

File: horqrux/circuit.py (hunk 1 above)
Comment (chMoussa) [reply horqrux/circuit.py]: @RolandMacDoland decided to add a field to separate feature map parameters from other variational parameters.

File: horqrux/matrices.py (hunk 3 above)
Comment (chMoussa) [reply horqrux/matrices.py:23]: @RolandMacDoland a `default_complex_dtype` should do the trick for allowing complex128 if the environment variable is set

File: horqrux/circuit.py [hunk 5]
```diff
@@ -1,47 +1,77 @@
...
+        """List of parameters of the circuit.
+            Composed of variational and feature map parameters.
+
+        Returns:
+            list[str]: Names of parameters.
+        """
+        return [str(op.param) for op in self.operations if isinstance(op, Parametric)]
+
+    @property
+    def variational_param_names(self) -> list[str]:
```
Comment (RolandMacDoland) [inline horqrux/circuit.py:53]:
```suggestion
    def vparam_names(self) -> list[str]:
```

Maybe a similar method for feature parameter names ?

**Review by RolandMacDoland — APPROVED** [review]
LGTM. Minor renaming quibble. I _think_ it is conventional in torch to have `vparams` and `fparams` to disambiguate.

File: horqrux/circuit.py (hunk 5 above)
Comment (chMoussa) [reply horqrux/circuit.py:53]: Gotcha, will do. Will just rename to `fparams` the input then and rename to `vparams` the method.

**Pull request merged**